internal/jobobject/jobobject.go

```diff
@@ -227,6 +227,25 @@ func (job *JobObject) SetIOLimit(maxBandwidth, maxIOPS int64) error {
 	return nil
 }
 
+// SetTerminateOnLastHandleClose sets the job object flag that specifies that the job should terminate
+// all processes in the job on the last open handle being closed.
+func (job *JobObject) SetTerminateOnLastHandleClose() error {
```

Comment:
Because they both set `JobObjectExtendedLimitInformation`, calling `SetMemoryLimit` and `SetTerminateOnLastHandleClose` will effectively blow away the setting made from the other one. I see a few ways to deal with this:

- Cache the most recently applied `JOBOBJECT_EXTENDED_LIMIT_INFORMATION` in the `JobObject`. Each time a new setting change is requested, update it and re-apply it.
- The same as above, but rather than caching, re-query with `QueryInformationJobObject`, update, then re-apply.
- Instead of providing separate functions like `SetMemoryLimits`, make this the caller's problem by just exposing a `SetExtendedLimits` function and making them pass in the `JOBOBJECT_EXTENDED_LIMIT_INFORMATION` struct themselves. This could be made more ergonomic with a wrapping type that supports a builder pattern:
  ```
      limits := jobobjects.NewExtendedLimits().SetMemoryLimit(100*1024*1024).SetTerminateOnLastHandleClose()
      if err := job.SetExtendedLimits(limits); err != nil {
      ...
  ```

I think #1 is probably the approach I prefer.

Reply:
Whatever we do, we should add a test that validates we can set both settings on a job object and query them back to ensure they are still set.

Reply:
1 sounds the best to me. I'm on it!

Reply:
I actually went with #2 as that's the direction Kathryn had gone with her recent job utility changes: https://github.com/microsoft/hcsshim/blob/c19ef4bd03a74f4ab40484963ad1d271a1820a0f/internal/jobobject/limits.go#L54

We can optimize this to just go off a cached version in another PR if that's alright. I don't want to do too much refactoring to the jobobject package in this change specifically.